### ISSUE TITLE
Shapes used to initialize variables must be fully-defined (no None dimensions).

### ISSUE THREAD

**Issue by RadulDev** [CityUHK-CompBio/DeepCC#19]
I was trying the tool with the given model, but came across this error. can any one help me guide me through using this tools correctly.

**Error:**
Shapes used to initialize variables must be fully-defined (no `None` dimensions). Received: shape=(None,) for variable path='sequential/batch_normalization/gamma'Traceback:

1. load_DeepCC_model("/home/dell/BRC_TCGA")
2. keras::load_model_hdf5(filepath = paste0(prefix, ".hdf5"))
3. load_model(filepath, custom_objects, compile)
4. do.call(keras$models$load_model, args)
5. (structure(function (filepath, custom_objects = NULL, compile = TRUE, 
 .     safe_mode = TRUE) 
 . {
 .     cl <- sys.call()
 .     cl[[1L]] <- list2
 .     call_args <- split_named_unnamed(eval(cl, parent.frame()))
 .     result <- py_call_impl(callable, call_args$unnamed, call_args$named)
 .     if (py_get_convert(callable)) 
 .         result <- py_to_r(result)
 .     if (is.null(result)) 
 .         invisible(result)
 .     else result
 . }, class = c("python.builtin.function", "python.builtin.object"
 . ), py_object = <environment>))(filepath = "/home/dell/BRC_TCGA.hdf5", 
 .     custom_objects = NULL, compile = TRUE)
6. py_call_impl(callable, call_args$unnamed, call_args$named)

**Comment by yaolei-leo** [comment]
I have the same issue. I downloaded the model from the baidunetdisk (Link: https://pan.baidu.com/s/10ogQPsNa4SksJadT8_ceHA      Password: ugeg) given by the author, and it shows the same issue:

**Error:**
Shapes used to initialize variables must be fully-defined (no None dimensions). Received: shape=(None,) for variable path='sequential/batch_normalization/gamma'Traceback:

It might because the given HDF5 files containing only the weights, while Keras requires both architecture and weights. I am still debugging.

**Comment by gaofeng21cn** [comment]
Fixed in 6b8d0fd.

The HDF5 files do contain the full architecture, so the cause is not missing architecture. Two entries in those files are rejected by Keras 3:

- The first layer stores `batch_input_shape`, which Keras 3 replaced with an explicit `Input` layer or `batch_shape`.
- `BatchNormalization` stores its axis as `[1]`. Keras 3 also changed the convention here: Keras 2 counted from the batch dimension, while Keras 3 counts feature axes, so the Keras 2 index has to be remapped to `-1`. Loading with the Keras 2 index is what produced the `shape=(None,)` error on `batch_normalization/gamma`.

`load_DeepCC_model()` now detects the file format and rebuilds Keras 2 models from their own recorded architecture, then restores the original weights.

Verified against the real published weights (`BRC_TCGA` and `CRC_TCGA`, pulled from Git LFS and SHA-256 checked): both load in about 3 seconds, produce 10-dimensional deep features, and return class probabilities that sum to 1. Weight loading is bit-exact against the HDF5 datasets. Loading a legacy model and then calling `save_DeepCC_model()` writes the current format and reproduces identical probabilities, so you only need the compatibility path once.

To pick up the fix:

```r
install.packages("remotes")
remotes::install_github("CityUHK-CompBio/DeepCC")
```

There is now a regression test that synthesises a Keras 2 HDF5 file and checks that loading it reproduces the equivalent Keras 3 predictions, so this stays fixed.
